File: MEMORY_TEST/JAVA/JIT_TEST/src/MAIN/SortingAlgorithm.java
package MAIN;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;

public abstract class SortingAlgorithm extends Algorithm {

    String sortDataPath;

    protected int[] random;
    protected int[] sortedAsc;
    protected int[] sortedDesc;
    protected int[] same;

    public SortingAlgorithm(String fileName, int iterations, String sortDataPath) {
        super(fileName, iterations);
        this.sortDataPath = sortDataPath;
    }

//    public SortingAlgorithm(String fileName, int iterations) {
//        super(fileName, iterations);
//    }

    protected PrintWriter prepareFileWriter(SortDataType type) {
        System.out.println(sortDataPath);
        try {
            String dataType = "";
            PrintWriter writer;
            switch ( type ) {
                case SAME: dataType += "SAME"; break;
                case RANDOM: dataType += "JIT_TEST"; break;
                case SORTED_ASC: dataType += "SORTED_ASC"; break;
                default: dataType += "SORTED_DESC"; break;
            }
            return new PrintWriter(filepath + fileName + dataType + getTimeAndDate() + ".txt", "UTF-8");
        } catch(Exception ex) {
            ex.printStackTrace();
        }
        return null;
    }

    protected void readDataFrom(SortDataType type) {

        BufferedReader br;
        try {
            String filename;
            switch(type) {
                case SAME: filename = "same.txt"; break;
                case RANDOM: filename = "random.txt"; break;
                case SORTED_ASC: filename = "sortedAsc.txt"; break;
                default: filename = "sortedDesc.txt"; break;
            }

            br = new BufferedReader(new FileReader(sortDataPath + filename));

            ArrayList<Integer> arr = new ArrayList<>();
            String line = br.readLine();

//            while (line != null && ctr <= Math.pow(2.0, 22)) {
            int ctr = 0;
            while ( line != null ) {
                arr.add(Integer.valueOf(line));
                line = br.readLine();
                ctr++;
            }
            br.close();

//            System.out.println(fileName);
//            for (int j = 0; j < arr.size(); j++) {
//                System.out.print(arr.get(j) + " ");
//                if(j % 50 == 0) System.out.println("");
//            }

            switch (type) {
                case RANDOM:
                    random = new int[arr.size()];
                    for (int i = 0; i < arr.size(); i++)
                        random[i] = arr.get(i);
                    return;
                case SAME:
                    same = new int[arr.size()];
                    for (int i = 0; i < arr.size(); i++)
                        same[i] = arr.get(i);
                    return;
                case SORTED_ASC:
                    sortedAsc = new int[arr.size()];
                    for (int i = 0; i < arr.size(); i++)
                        sortedAsc[i] = arr.get(i);
                    return;
                default: //sortedDesc
                    sortedDesc = new int[arr.size()];
                    for (int i = 0; i < arr.size(); i++)
                        sortedDesc[i] = arr.get(i);
                    return;
            }

        } catch(Exception ex) {
            ex.printStackTrace();
        }

    }




}

File: MEMORY_TEST/CPP/MERGE_MEM/SortingAlgorithm.cpp
//
// Created by vm on 02.11.17.
//

#include <vector>
#include <algorithm>
#include "SortingAlgorithm.h"


std::fstream SortingAlgorithm :: prepareFileWriter(SortDataType type) {
    std::string dataType;
    switch(type) {
        case SAME : dataType = "SAME"; break;
        case RANDOM : dataType = "RANDOM"; break;
        case SORTED_ASC : dataType = "SORTED_ASC"; break;
        default: dataType = "SORTED_DESC"; break;
    }
    std::fstream fs;
    fs.open(filePath + fileName + dataType + getTimeAndDate(), std::fstream::out);
    //std::cout << fs.is_open() << std::endl;
    return fs;
}

void SortingAlgorithm :: readDataFrom(SortDataType  type ) {
    std::string filename;
    switch(type) {
        case SAME: filename = "same.txt"; break;
        case RANDOM: filename = "random.txt"; break;
        case SORTED_ASC: filename = "sortedAsc.txt"; break;
        default: filename = "sortedDesc.txt"; break;
    }
    std::ifstream reader;
    reader.open(sortDataPath + filename);
    std::vector<int> arr;// = std::vector<int>();
    char line[100];
    if(reader.is_open()) {
        while(!reader.eof()) {
            reader >> line;
            //std::cout << std::stoi(line) << std::endl;
            arr.push_back( std::stoi(line) ); // string to int conversion
        }
    }
    reader.close();
    switch(type) {
        case RANDOM:
            random.swap(arr);
            arr.shrink_to_fit();
            return;
        case SAME:
            same.swap(arr);
            arr.shrink_to_fit();
            return;
        case SORTED_ASC:
            sortedAsc.swap(arr);
            arr.shrink_to_fit();
            return;
        default: //sortedDesc
            sortedDesc.swap(arr);
            arr.shrink_to_fit();
            return;
    }
}


//    int vals[size];
//    std::copy(arr.begin(), arr.end(), vals);
//    switch (type) {
//        case RANDOM:
//            random = vals;
//            arr.shrink_to_fit();
//            return;
//        case SAME:
//            same = vals;
//            arr.shrink_to_fit();
//            return;
//        case SORTED_ASC:
//            sortedAsc = vals;
//            arr.shrink_to_fit();
//            return;
//        default: //sortedDesc
//            sortedDesc = vals;
//            arr.shrink_to_fit();
//            return;
//    }





